extraction/src/main/scala/org/broadinstitute/monster/encode/extraction/ExtractionPipelineBuilder.scala

```diff
@@ -24,52 +21,86 @@ import scala.concurrent.Future
 class ExtractionPipelineBuilder(getClient: () => EncodeClient)
     extends PipelineBuilder[Args]
     with Serializable {
+  import ExtractionPipelineBuilder._
+
   implicit val coder: Coder[Msg] = Coder.beam(new UpackMsgCoder)
 
+  /** Convenience alias to make signatures a little less nasty. */
+  private type LookupParams = (EncodeEntity, List[(String, String)])
+
+  private val lookupFn =
+    new ScalaAsyncLookupDoFn[LookupParams, Msg, EncodeClient](MaxConcurrentRequests) {
+
+      override def asyncLookup(
+        client: EncodeClient,
+        params: LookupParams
+      ): Future[Msg] = {
+        val (encodeEntity, filters) = params
+        client.get(encodeEntity, filters)
+      }
+      override protected def newClient(): EncodeClient = getClient()
+    }
+
+  /**
+    * Map a  batch of search parameters into MessagePack entities from
+    * ENCODE matching those parameters.
+    *
+    * @param encodeEntity the type of ENCODE entity the stage should query
+    * @param filterBatches batches of key=value filters to include in API queries
+    */
+  def getEntities(
+    encodeEntity: EncodeEntity,
+    filterBatches: SCollection[List[(String, String)]]
+  ): SCollection[Msg] =
+    filterBatches
+      .withName("Construct full query parameters")
+      .map(encodeEntity -> _)
+      .withName("Query ENCODE API")
+      .applyKvTransform(ParDo.of(lookupFn))
+      .withName("Extract result graph")
+      .flatMap(_.getValue.fold(throw _, _.read[Array[Msg]]("@graph")))
```

Comment:
Moved this up to be closer to the `lookupFn` definition, and added `withName`s so each step is clear in the UI.